conda/gateways/disk/delete.py

```diff
@@ -27,35 +27,39 @@
 
 def rmtree(path, *args, **kwargs):
     # subprocessing to delete large folders can be quite a bit faster
+    path = normpath(path)
     if on_win:
         try:
-            check_call('rd /s /q "{}"'.format(path), shell=True)
-        except CalledProcessError as e:
-            if e.returncode == 5:
-                # skip permission errors
-                log.warn("")
-                pass
-            else:
-                raise
+            # the fastest way seems to be using DEL to recursively delete files
+            # https://www.ghacks.net/2017/07/18/how-to-delete-large-folders-in-windows-super-fast/
+            # However, this is not entirely safe, as it can end up following symlinks to folders
+            # https://superuser.com/a/306618/184799
+            # so, we stick with the slower, but hopefully safer way.  Maybe if we figured out how
+            #    to scan for any possible symlinks, we could do the faster way.
+            # out = check_output('DEL /F/Q/S *.* > NUL 2> NUL'.format(path), shell=True,
+            #                    stderr=STDOUT, cwd=path)
+            out = check_output('RD /S /Q "{}" > NUL 2> NUL'.format(path), shell=True,
+                               stderr=STDOUT)
+        except CalledProcessError:
+            log.debug("removing dir contents the fast way failed.  Output was: {}".format(out))
     else:
         try:
             makedirs('.empty')
         except:
             pass
         # yes, this looks strange.  See
         #    https://unix.stackexchange.com/a/79656/34459
-        #    https://web.archive.org/web/20130929001850/http://linuxnote.net/jianingy/en/linux/a-fast-way-to-remove-huge-number-of-files.html
+        #    https://web.archive.org/web/20130929001850/http://linuxnote.net/jianingy/en/linux/a-fast-way-to-remove-huge-number-of-files.html  # NOQA
         rsync = which('rsync')
-        if rsync:
+        if rsync and isdir('.empty'):
             try:
-                check_call([rsync, '-a', '--delete', join(getcwd(), '.empty') + "/", path + "/"])
+                out = check_output(
+                    [rsync, '-a', '--delete', join(getcwd(), '.empty') + "/", path + "/"],
+                    stderr=STDOUT)
             except CalledProcessError:
-                log.warn("Removing folder {} the fast way (with rsync) failed.  "
-                         "Do you have rsync available?".format(path))
-                shutil.rmtree(path)
-        else:
-            shutil.rmtree(path)
-        shutil.rmtree('.empty')
+                log.debug("removing dir contents the fast way failed.  Output was: {}".format(out))
+            shutil.rmtree('.empty')
+    shutil.rmtree(path)
```

Comment:
The call to shutil.rmtree() is unconditional now.

Reply:
I think that's OK.  It's sort of like a final "if this is still really around for any reason, try one more time with Python to clean it up."